packages/dd-trace/src/debugger/devtools_client/condition.js

```diff
@@ -0,0 +1,237 @@
+      case 'endsWith': return `String.prototype.endsWith.call(${assertString(args[0])}, ${assertString(args[1])})`
+      case 'contains': return `((obj, elm) => {
+          if (${isString('obj')}) {
+            return String.prototype.includes.call(obj, elm)
+          } else if (Array.isArray(obj)) {
+            return Array.prototype.includes.call(obj, elm)
+          } else if (obj instanceof Object.getPrototypeOf(Int8Array)) {
+            return Object.getPrototypeOf(Int8Array.prototype).includes.call(obj, elm)
+          } else if (obj instanceof Set) {
+            return Set.prototype.has.call(obj, elm)
+          } else if (obj instanceof WeakSet) {
+            return WeakSet.prototype.has.call(obj, elm)
+          } else if (obj instanceof Map) {
+            return Map.prototype.has.call(obj, elm)
+          } else if (obj instanceof WeakMap) {
+            return WeakMap.prototype.has.call(obj, elm)
+          } else {
+            throw new TypeError('Variable does not support contains')
+          }
+        })(${args[0]}, ${args[1]})`
+      case 'matches': return `((str, regex) => {
+          if (${isString('str')}) {
+            if (Object.getPrototypeOf(regex) === RegExp.prototype) {
+              return RegExp.prototype.test.call(regex, str)
+            } else if (${isString('regex')}) {
+              return String.prototype.match.call(str, regex) !== null
+            } else {
+              throw new TypeError('Regular expression must be either a string or an instance of RegExp')
+            }
+          } else {
+            throw new TypeError('Variable is not a string')
+          }
+        })(${args[0]}, ${args[1]})`
+      case 'filter': return `(($dd_var) => {
+          return ${isIterableCollection('$dd_var')}
+            ? Array.from($dd_var).filter(($dd_it) => ${args[1]})
+            : Object.entries($dd_var).filter(([$dd_key, $dd_value]) => ${args[1]}).reduce((acc, [k, v]) => {
+                acc[k] = v
+                return acc
+              }, {})
+        })(${args[0]})`
+      case 'substring': return `((str) => {
+          if (${isString('str')}) {
+            return String.prototype.substring.call(str, ${args[1]}, ${args[2]})
+          } else {
+            throw new TypeError('Variable is not a string')
+          }
+        })(${args[0]})`
+      case 'getmember': return accessProperty(args[0], args[1], false)
+      case 'index': return accessProperty(args[0], args[1], true)
+    }
+  }
+
+  throw new TypeError(`Unknown AST node type: ${type}`)
+}
+
+function iterateOn (fnName, variable, callbackCode) {
+  return `(($dd_val) => {
+    return ${isIterableCollection('$dd_val')}
+      ? Array.from($dd_val).${fnName}(($dd_it) => ${callbackCode})
+      : Object.entries($dd_val).${fnName}(([$dd_key, $dd_value]) => ${callbackCode})
+  })(${variable})`
+}
+
+function isString (variable) {
+  return `typeof ${variable} === 'string' || ${variable} instanceof String`
```

Comment:
Do we have to handle boxed primitives?

Reply:
I agree it's an edge case, but you'll be surprised what kind of code people write in the wild. You could of course argue that we just don't support that type of strings, but I guess it's an easy win with this extra `instanceof` check, so I figured, why not 🤷